File: scripts/foundry/update/smoke.py
#!/usr/bin/env python3
"""Post-update verification, and the honest definition of "it still works".

Two layers, because they catch different things and only one of them is cheap.

**Server side** proves the world launches and the server is not throwing: the world
becomes active with the expected system, nothing new lands in Foundry's error log, and
no package we touched is in ``packageWarnings``.

**Client side** proves the part the server cannot see. A module that throws in the
browser and leaves the canvas blank is, from the server's point of view, a completely
healthy world. That failure would be discovered at the table.

Two things about the sequencing matter more than the checks themselves:

1. **Launching a world is itself destructive.** ``world.mjs setup()`` runs
   ``migrateCore()`` when the core is newer and ``migrateSystem()`` when the system is,
   in place, across every document, before anything here can pass or fail. That is why
   the backups happen first and why a failure at this stage restores *data*, not just code.

2. **Every world migrates independently, when it is launched.** Smoking only
   ``space-journey`` would leave ``ardenhaven`` to migrate unobserved at Joe's next
   manual launch — with no fresh snapshot and nobody watching. So all worlds are
   smoked, sequentially, and the report records which ones migrated.
"""

from __future__ import annotations

import json
import re
import subprocess
import tempfile
import time
from dataclasses import dataclass, field
from pathlib import Path

from scripts.foundry.update import admin as admin_mod
from scripts.foundry.update.admin import FoundryAdmin, Timeout

REPO_ROOT = Path(__file__).resolve().parents[3]
SMOKE_JS = REPO_ROOT / "automation" / "smoke" / "smoke.mjs"
NODE = "/opt/homebrew/bin/node"
CHROME_PROFILE = REPO_ROOT / ".state" / "smoke-chrome-profile"
BASELINE_PATH = REPO_ROOT / ".state" / "vtt-smoke-baseline.json"
# Time for a deactivated world's LevelDB to finish closing before the next launch.
SETTLE_SECONDS = 8.0

_DIGITS = re.compile(r"\d+")

# Modules whose warnings are noise rather than breakage.
IGNORED_WARNING_TYPES = {"info"}


@dataclass
class WorldResult:
    world: str
    ok: bool = False
    server_ok: bool = False
    client_ok: bool | None = None       # None when the browser check was skipped
    migrated_system: str | None = None
    migrated_core: str | None = None
    errors: list[str] = field(default_factory=list)
    notes: list[str] = field(default_factory=list)
    client: dict = field(default_factory=dict)

    def to_json(self) -> dict:
        return {
            "world": self.world, "ok": self.ok, "server_ok": self.server_ok,
            "client_ok": self.client_ok, "migrated_system": self.migrated_system,
            "migrated_core": self.migrated_core, "errors": self.errors,
            "notes": self.notes,
            # The full client payload carries console text; keep the useful summary.
            "client": {k: self.client.get(k) for k in
                       ("gameReady", "canvasReady", "scene", "systemVersion",
                        "coreVersion", "durationMs", "loggedInAs", "note")},
        }


def run_all(fa: FoundryAdmin, worlds: list[dict], *, policy: dict,
            touched: set[str], error_offset: int,
            log_day: str | None = None) -> list[WorldResult]:
    """Smoke every world in turn, leaving the last one launched."""
    smoke_policy = policy.get("smoke", {}) or {}
    wanted = smoke_policy.get("worlds", "all")
    selected = [w for w in worlds
                if wanted == "all" or w["id"] in (wanted if isinstance(wanted, list) else [wanted])]

    results: list[WorldResult] = []
    for world in selected:
        # Foundry serves ONE world at a time, and `launchWorld` is a /setup action —
        # which 403s while any world is active. So the previous world has to come down
        # before the next can go up, or every world after the first fails to launch.
        _return_to_setup(fa)
        results.append(_smoke_one(fa, world, policy=policy, touched=touched,
                                  error_offset=error_offset, log_day=log_day))
        # Each world's launch appends to the same log; advance the mark so the next
        # world is not blamed for the previous one's output.
        error_offset = admin_mod.error_log_offset(log_day)
    return results


def _return_to_setup(fa: FoundryAdmin) -> None:
    """Deactivate whatever world is running, so the setup API answers again.

    Prefers the graceful path (`POST /join {shutdown}`, which needs the admin password)
    and falls back to restarting the app — `options.json` has `world: null`, so a fresh
    start lands on the setup screen.
    """
    try:
        if not fa.world_active():
            return
    except Exception:  # noqa: BLE001
        return
    if fa.deactivate_world():
        # /api/status flips to inactive as soon as the world is torn down, but the
        # world's LevelDB closes asynchronously after that. Launching the next world
        # too quickly produces `LEVEL_DATABASE_NOT_OPEN — Failed to connect to database`
        # in the server log, which the error-log check then reports as breakage caused
        # by the update. Observed; the settle is not superstition.
        time.sleep(SETTLE_SECONDS)
        return
    admin_mod.quit_app()
    admin_mod.start_app()
    try:
        fa.reauthenticate()   # the restart invalidated the session
    except Exception:  # noqa: BLE001
        pass


def _fingerprint(errors: list[dict]) -> list[str]:
    """A stable signature for a world's client errors.

    Numbers are masked because line/column offsets and ids move between builds while
    the underlying fault is the same one.
    """
    seen = {_DIGITS.sub("#", (e.get("text") or "")[:200]).strip()
            for e in errors if e.get("type") in ("pageerror", "console", "canvas")}
    return sorted(x for x in seen if x)


def load_baseline() -> dict:
    try:
        return json.loads(BASELINE_PATH.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return {}


def save_baseline(world: str, fingerprint: list[str]) -> None:
    BASELINE_PATH.parent.mkdir(parents=True, exist_ok=True)
    data = load_baseline()
    data[world] = fingerprint
    BASELINE_PATH.write_text(json.dumps(data, indent=2, sort_keys=True), encoding="utf-8")


def _smoke_one(fa: FoundryAdmin, world: dict, *, policy: dict, touched: set[str],
               error_offset: int, log_day: str | None) -> WorldResult:
    smoke_policy = policy.get("smoke", {}) or {}
    result = WorldResult(world=world["id"])

    pre_system = world.get("system_version")
    pre_core = world.get("core_version")

    # ── Server side ──
    try:
        fa.launch_world(world["id"])
        status = fa.wait_for_world(world["id"], timeout=240)
    except (Timeout, Exception) as exc:  # noqa: BLE001
        result.errors.append(f"world did not launch: {exc}")
        return result

    if status.system_version and status.system_version != pre_system:
        result.migrated_system = f"{pre_system} → {status.system_version}"
    if status.version and pre_core and status.version != pre_core:
        result.migrated_core = f"{pre_core} → {status.version}"

    new_errors = admin_mod.error_log_since(error_offset, log_day).strip()
    if new_errors:
        # Keep a bounded excerpt: the log can contain player IPs, and the report is
        # committed, so report.py scrubs this before it is written out.
        result.errors.append("new server errors:\n" + new_errors[-2000:])

    result.server_ok = not result.errors

    # ── Client side ──
    if not smoke_policy.get("browser", True):
        result.notes.append("browser check disabled in update-policy.yml")
        result.ok = result.server_ok
        return result

    client = run_browser(fa, world["id"],
                         dwell=int(smoke_policy.get("browser_dwell_seconds", 60)),
                         ignores=smoke_policy.get("ignore_console") or [])
    result.client = client

    # A world that was ALREADY throwing before we touched anything must not veto an
    # update. `ardenhaven` throws "Cannot add property walls, object is not extensible"
    # on every load, entirely independently of this run — without a baseline, that one
    # broken world would roll back every good update forever. So the pass/fail question
    # is "did this run make it worse", not "is it perfect".
    fingerprint = _fingerprint(client.get("errors", []))
    baseline = load_baseline().get(world["id"])
    new_faults = ([f for f in fingerprint if f not in baseline]
                  if baseline is not None else [])

    if baseline is None:
        # First sighting: nothing to compare against, so record rather than accuse.
        result.client_ok = True
        if fingerprint:
            result.notes.append(
                f"{len(fingerprint)} pre-existing client error(s) recorded as this "
                f"world's baseline; a future run fails only on NEW ones")
        save_baseline(world["id"], fingerprint)
    elif new_faults:
        result.client_ok = False
        for fault in new_faults[:10]:
            result.errors.append(f"client (new since last run): {fault}")
    else:
        result.client_ok = True
        if fingerprint:
            result.notes.append(f"{len(fingerprint)} known client error(s), unchanged "
                                f"from the recorded baseline")
        # Refresh, so a fault that has since been FIXED stops being tolerated.
        if fingerprint != baseline:
            save_baseline(world["id"], fingerprint)
    if not client.get("gameReady"):
        # Whatever the baseline says, a world that never becomes ready is a failure.
        result.client_ok = False
        result.errors.append("client: game.ready never became true")
    if client.get("note"):
        result.notes.append(client["note"])

    # A module that was updated and now warns is the signal we care about; unrelated
    # pre-existing noise is not this run's problem.
    for module in client.get("activeModules", []):
        pkg_id = module.split("@")[0]
        if pkg_id in touched:
            result.notes.append(f"updated module active after reload: {module}")

    result.ok = result.server_ok and result.client_ok
    return result


def run_browser(fa: FoundryAdmin, world: str, *, dwell: int,
                ignores: list[str]) -> dict:
    """Drive Chrome through the world as GM and report what the browser saw.

    The admin session cookie is handed to Node through a 0600 temp file rather than an
    argument — cookies are bearer credentials and argv is world-readable. The GM's own
    password is never involved: the browser is already an authenticated admin and uses
    ``loginAs``.
    """
    if not SMOKE_JS.exists():
        return {"ok": True, "skipped": "smoke.mjs not installed", "errors": []}

    try:
        fa.authenticate()
    except Exception as exc:  # noqa: BLE001
        return {"ok": False, "errors": [{"type": "fatal",
                                         "text": f"admin auth failed: {exc}"}]}

    cookies = fa.session_cookies()
    if not cookies:
        return {"ok": False, "errors": [{"type": "fatal",
                                         "text": "no admin session cookie to hand the browser"}]}

    CHROME_PROFILE.mkdir(parents=True, exist_ok=True)
    with tempfile.NamedTemporaryFile("w", suffix=".json", delete=False) as fh:
        json.dump(cookies, fh)
        cookie_path = Path(fh.name)
    cookie_path.chmod(0o600)

    cmd = [NODE, str(SMOKE_JS), "--url", fa.base_url, "--world", world,
           "--dwell", str(dwell), "--cookies", str(cookie_path),
           "--profile", str(CHROME_PROFILE)]
    for pattern in ignores:
        cmd += ["--ignore", pattern]

    try:
        proc = subprocess.run(cmd, capture_output=True, text=True,
                              timeout=dwell + 600)
    except (subprocess.TimeoutExpired, OSError) as exc:
        return {"ok": False, "errors": [{"type": "fatal",
                                         "text": f"browser smoke test failed to run: {exc}"}]}
    finally:
        cookie_path.unlink(missing_ok=True)

    try:
        return json.loads(proc.stdout[proc.stdout.find("{"):])
    except (json.JSONDecodeError, ValueError):
        return {"ok": False, "errors": [
            {"type": "fatal",
             "text": f"unparseable smoke output (rc={proc.returncode}): "
                     f"{(proc.stderr or proc.stdout)[-500:]}"}]}


def package_warnings(fa: FoundryAdmin, touched: set[str]) -> dict:
    """Foundry's own per-package warnings.

    Must be read **before** a world is launched: ``setup.mjs``'s socket handler returns
    ``{}`` once ``game.world`` is set, so this is silently empty afterwards. It is also
    the only place an install-step failure shows up — ``installPackage`` resolves ``{}``
    on one.
    """
    try:
        data = fa.setup("getPackages", type="module")
    except Exception:  # noqa: BLE001
        return {}
    warnings = data.get("packageWarnings") or {}
    return {k: v for k, v in warnings.items() if k in touched}


def wait_for_settle(seconds: float = 5.0) -> None:
    """Let the server finish its post-launch bookkeeping before probing again."""
    time.sleep(seconds)

File: automation/smoke/smoke.mjs
/**
 * Client-side smoke test for a Foundry world, run after an unattended update.
 *
 * The server-side checks (world launches, no new errors in the server log, no package
 * warnings) are blind to the failure that actually ruins a session: a module that
 * throws in the browser and leaves the canvas blank. The server is perfectly happy.
 * This is the cheapest honest way to catch that.
 *
 * It is also what forces dnd5e's *client-side* world migration to happen now, inside
 * the window where the backups are fresh and a rollback is one command — rather than at
 * the GM's next login, days later, with no snapshot taken since.
 *
 * Login needs no *user* password. The caller (smoke.py) has already authenticated as
 * server admin over 127.0.0.1 and hands the session cookie in on a 0600 temp file;
 * `sessions.loginAsUser` then permits logging in as any user without that user's own
 * password. It genuinely requires the admin session: `authenticateAdmin` returns
 * success without setting `session.admin` when no admin password is configured, so
 * this path 403s (`USERS.LoginAsGMRequired`) unless one is set. The cookie is passed
 * by file, not argument — it is a bearer credential and argv is world-readable.
 *
 * Because the Cloudflare tunnel 403s /auth and /setup, admin is a localhost-only
 * capability by construction, which is the "admin only on the local port" model this
 * setup wants.
 *
 * Chrome runs headful but positioned off-screen: Foundry's canvas is WebGL, headless
 * SwiftShader is a different renderer than the one Joe's players use, and a LaunchAgent
 * runs in the GUI session so a real window is available. Testing on the same renderer
 * the table uses is the point.
 *
 * Usage:
 *   node smoke.mjs --url http://127.0.0.1:30000 --world space-journey \
 *                  --dwell 60 --out result.json [--ignore "substr" ...]
 * Exit code is always 0; the verdict is in the JSON. A crash here is a smoke-test
 * failure, not an update failure, and the caller decides what that is worth.
 */

import { createRequire } from "node:module";
import { writeFileSync, readFileSync } from "node:fs";

const require = createRequire(import.meta.url);
const puppeteer = require("puppeteer-core");

const CHROME = "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome";
const GAMEMASTER_ROLE = 4;

function arg(name, fallback = null) {
  const i = process.argv.indexOf(`--${name}`);
  return i > -1 && process.argv[i + 1] ? process.argv[i + 1] : fallback;
}
function argAll(name) {
  const out = [];
  process.argv.forEach((a, i) => { if (a === `--${name}`) out.push(process.argv[i + 1]); });
  return out;
}

const baseUrl = arg("url", "http://127.0.0.1:30000").replace(/\/$/, "");
const world = arg("world", null);
const dwellMs = Number(arg("dwell", "60")) * 1000;
const outPath = arg("out", null);
const ignores = argAll("ignore");
const profileDir = arg("profile", "/tmp/vtt-smoke-chrome");

const result = {
  ok: false, world, stage: "start", errors: [], warnings: [],
  gameReady: false, canvasReady: false, activeModules: [], scene: null,
  durationMs: 0, note: null,
};

const started = Date.now();
let browser;

const ignored = (text) => ignores.some((p) => p && text.includes(p));

try {
  browser = await puppeteer.launch({
    executablePath: CHROME,
    headless: false,
    // Off-screen rather than headless: same renderer as the players, no window in the
    // way if Joe happens to be at the machine.
    args: [
      "--window-position=-4000,-4000", "--window-size=1600,1000",
      "--no-first-run", "--no-default-browser-check",
      "--disable-features=Translate,MediaRouter",
    ],
    userDataDir: profileDir,
    protocolTimeout: 300000,
  });

  const page = await browser.newPage();
  await page.setViewport({ width: 1600, height: 1000 });

  // The admin session, minted by smoke.py over 127.0.0.1 and handed over by file.
  const cookiePath = arg("cookies", null);
  if (cookiePath) {
    const cookies = JSON.parse(readFileSync(cookiePath, "utf8"));
    if (cookies.length) {
      await browser.setCookie(...cookies.map((c) => ({ ...c, url: baseUrl })));
    }
  }

  page.on("pageerror", (err) => {
    const text = String(err?.message || err);
    if (!ignored(text)) result.errors.push({ type: "pageerror", text: text.slice(0, 800) });
  });
  page.on("console", (msg) => {
    const text = msg.text();
    if (ignored(text)) return;
    if (msg.type() === "error") result.errors.push({ type: "console", text: text.slice(0, 800) });
    else if (msg.type() === "warning") result.warnings.push(text.slice(0, 400));
  });

  // 1. Land on /join so the page's own client data (game.users, with roles) is available.
  result.stage = "join";
  await page.goto(`${baseUrl}/join`, { waitUntil: "networkidle2", timeout: 120000 });
  await page.waitForFunction("typeof game !== 'undefined' && !!game.users", { timeout: 60000 });

  const users = await page.evaluate(() =>
    game.users.map((u) => ({ id: u._id || u.id, name: u.name, role: u.role })));
  const gm = users.find((u) => u.role === GAMEMASTER_ROLE) || users[0];
  if (!gm) throw new Error("no users in this world to log in as");
  result.loggedInAs = gm.name;

  // 2. Log in as the GM. The session is already admin via the cookie, so this
  //    needs no user password.
  result.stage = "login";
  const login = await page.evaluate(async (userId) => {
    // No /auth call here: the session cookie already carries admin. Posting /auth
    // again with an empty body would only re-run the password check and fail.
    const r = await fetch("/join", {
      method: "POST", credentials: "same-origin",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ action: "loginAs", userId }),
    });
    return { status: r.status, body: await r.text() };
  }, gm.id);
  if (login.status !== 200) {
    throw new Error(`loginAs failed: HTTP ${login.status} ${login.body.slice(0, 200)}`);
  }

  // 3. The real test. A world with a broken module usually dies between `game.ready`
  //    and `canvas.ready`, so both are checked, separately.
  result.stage = "game";
  await page.goto(`${baseUrl}/game`, { waitUntil: "domcontentloaded", timeout: 180000 });

  await page.waitForFunction("typeof game !== 'undefined' && game.ready === true",
    { timeout: 240000, polling: 1000 });
  result.gameReady = true;

  try {
    await page.waitForFunction("typeof canvas !== 'undefined' && canvas.ready === true",
      { timeout: 120000, polling: 1000 });
    result.canvasReady = true;
  } catch {
    // A world whose every scene is unset has no canvas to draw. That is a legitimate
    // state, not a failure — record it rather than failing the update over it.
    const anyScene = await page.evaluate(() => !!game.scenes?.size).catch(() => false);
    result.note = anyScene
      ? "canvas never became ready despite the world having scenes"
      : "no active scene — canvas readiness not applicable";
    if (anyScene) result.errors.push({ type: "canvas", text: result.note });
  }

  // 4. Sit there. Module errors frequently arrive well after ready, on the first
  //    render tick or the first hook that fires.
  result.stage = "dwell";
  await new Promise((r) => setTimeout(r, dwellMs));

  Object.assign(result, await page.evaluate(() => ({
    activeModules: game.modules.filter((m) => m.active).map((m) => `${m.id}@${m.version}`),
    scene: game.scenes?.active?.name ?? null,
    systemVersion: game.system?.version ?? null,
    coreVersion: game.version ?? game.release?.version ?? null,
  })));

  result.stage = "done";
  result.ok = result.errors.length === 0;
} catch (err) {
  result.errors.push({ type: "fatal", text: String(err?.message || err).slice(0, 1000) });
  result.ok = false;
} finally {
  result.durationMs = Date.now() - started;
  if (browser) { try { await browser.close(); } catch { /* already gone */ } }
}

const json = JSON.stringify(result, null, 2);
if (outPath) writeFileSync(outPath, json);
process.stdout.write(json + "\n");
